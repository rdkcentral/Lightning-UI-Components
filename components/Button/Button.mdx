import { Preview, Story } from '@storybook/addon-docs/blocks';
// uncomment the line below and edit the path to point to your component
// import * as stories from './Button.stories';

# Button

Basic component that handles styling and functionality for a button

## Usage

```js
import { Button } from '@lightning/ui';

class Example extends lng.Component {
  static _template() {
    return {
      Button: {
        type: Button,
        title: 'Button'
      }
    };
  }
};
```

Before the button receives its title, it should be in a loading state.

<Preview>
  <Story id="button--basic" />
</Preview>


### Sizing

Buttons can take in a`w`. This is the minimum width that a button will be, the button will dynamically size itself based on the width of the title provided. Buttons can also take in a `h` - if not provided it will default to `40px`.

### Focus

When focused, the background/foreground colors are inverted by default. This can be overridden with theming.

### Icon

All buttons have the ability to be passed an icon. This will position the icon to the left of the title and center it horizontally and vertically.

### Theme

Themes have the ability to override default style properties for button mount and focus and unfocus states.
Focus and unfocus can take in patch functions that update styles on button _focus and _unfocus. If a theme property is also set directly on the Button component, the property set on the component itself will override the theme.

```js
  radius: 0,
  text: { fontSize: 20 },
  w: 150,
  h: 40,
  padding: 50,
  unfocus: {
    background: 0xff1f1f1f,
    text: 0xffffffff,
    icon: 0xffffffff,
    patch: function() {
      this.patch({
        smooth: { color: this._theme.unfocus.background },
        Content: {
          Title: { smooth: { color: this._theme.unfocus.text } },
          Icon: { smooth: { color: this._theme.unfocus.icon } }
        }
      });
    }
  },
  focus: {
    background: 0xffffffff,
    radius: 0,
    text: 0xff1f1f1f,
    icon: 0xff1f1f1f,
    patch: function() {
      this.patch({
        smooth: { color: this._theme.focus.background },
        Content: {
          Title: { smooth: { color: 0xff1f1f1f } },
          Icon: { smooth: { color: this._theme.focus.icon } }
        }
      });
    }
  }
}
```




## API

### Properties

name|type|readonly|description
--|--|--|--
announce|string|true|announcement text for screen readers
h|number|false|height of the button
w|number|false|width of the button
title|string|false|button text
radius|rumber|false|corner radius for a button
stroke|object|false|`color` and `weight` properties for button stroke
loading|object|false|`tag` and `animation` properties for button loading animation
icon|object|false|`src`, `size`, and `spacing` properties for an icon
theme|object|false|properties that apply style to the button

### Methods

#### _handleEnter

Can be overridden to apply handling for the <kbd>Enter</kbd> key
