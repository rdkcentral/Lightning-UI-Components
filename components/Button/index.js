/**
 * Button Component
 *
 * Generic Button Component that handles setting text
 * and focus and unfocus states for the button background.
 */
import lng from 'wpe-lightning';
import { RoundRect } from '../../utils';
import withStyles from '../../mixins/withStyles';
import Icon from '../Icon';

export const styles = {
  w: 150,
  h: 40,
  radius: 0,
  background: { color: 0xff1f1f1f },
  icon: { color: 0xffffffff },
  text: {
    fontSize: 20,
    color: 0xffffffff
  },
  padding: 50,
  unfocus: {
    patch: function() {
      const { background, icon, text } = this.styles;
      this.patch({
        smooth: { color: background.color },
        Content: {
          Title: { smooth: { color: text.color } },
          Icon: { smooth: { color: icon.color } }
        }
      });
    }
  },
  focus: {
    background: 0xffffffff,
    text: 0xff1f1f1f,
    icon: 0xff1f1f1f,
    patch: function() {
      const { background, icon, text } = this.styles.focus;
      this.patch({
        smooth: { color: background },
        Content: {
          Title: { smooth: { color: text } },
          Icon: { smooth: { color: icon } }
        }
      });
    }
  }
};

class Button extends lng.Component {
  static _template() {
    return {
      w: this.styles.w,
      h: this.styles.h,
      radius: this.styles.radius,
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2,
        flex: {
          direction: 'row',
          alignContent: 'center',
          alignItems: 'center'
        },
        Icon: {
          type: Icon
        },
        // TODO: Wonky lineHeight / fontsize from Lightning
        // Move title down 2 pixels to _visually_ center it
        // inside the button
        Title: { y: 2 }
      },
      Stroke: {
        zIndex: -1,
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2
      }
    };
  }

  _init() {
    this._update();
  }

  _focus() {
    this.styles.focus.patch.apply(this);
  }

  _unfocus() {
    this.styles.unfocus.patch.apply(this);
  }

  _update() {
    const template = { Content: {} };
    const Title = {};
    const Icon = {};
    const Stroke = {};

    const radius = this.radius || this.styles.radius;

    template.color = [this.background, this.styles.background.color].find(
      Number.isFinite
    );

    if (this.title) {
      Title.text = {
        ...this.styles.text,
        fontColor: this.styles.text.color,
        fontSize: this.fontSize || this.styles.text.fontSize,
        text: this.title
      };
      Title.color = this.styles.text.color;
      template.Content.Title = Title;
    }

    if (this.icon) {
      const { color, size, spacing, src } = this.icon;
      Icon.color = color || this.styles.icon.color;
      Icon.w = size;
      Icon.h = size;
      Icon.flexItem = { marginRight: spacing };
      Icon.icon = src;
      template.Content.Icon = Icon;
    }

    if (this.stroke) {
      const { color, weight } = this.stroke;
      const radius = this.radius || this.styles.radius;

      template.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(this.w),
        RoundRect.getHeight(this.h),
        radius,
        0x00,
        true,
        0xffffffff
      );

      Stroke.color = color;
      Stroke.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(this.w),
        RoundRect.getHeight(this.h),
        radius,
        weight,
        0xffffffff,
        true,
        this.background
      );

      template.Stroke = Stroke;
    } else {
      template.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(this.w),
        RoundRect.getHeight(this.h),
        radius
      );
    }

    this.patch(template);

    this._Title.on('txLoaded', () => {
      let iconSize = this._icon ? this._icon.size + this._icon.spacing : 0;
      let padding = [this.padding, this.styles.padding, 50].find(
        Number.isFinite
      );
      if (!this.fixed) {
        const w = this._Title.renderWidth + padding * 2 + iconSize;
        if (w !== this.w) {
          this.w = w > this.w ? w : this.w;
          this.fireAncestors('$itemChanged');
          this.signal('buttonWidthChanged', { w: this.w });
          this._update();
        }
      }
    });
  }

  _handleEnter() {
    if (typeof this.onEnter === 'function') {
      this.onEnter(this);
    }
  }

  get title() {
    return this._title;
  }

  set title(title) {
    this._title = title;
    this._update();
  }

  get icon() {
    return this._icon;
  }

  set icon({ src, size = 20, spacing = 5, color = 0xffffffff }) {
    if (src) {
      this._icon = { src, size, spacing, color };
      this._update();
    }
  }

  get stroke() {
    return this._stroke || this.styles.stroke;
  }

  set stroke({ weight = 2, color = 0x00 }) {
    this._stroke = { weight, color };
    this._update();
  }

  get announce() {
    // TODO - Localization?
    // Do we need a locale file with
    // component translations?
    return this._title + ', Button';
  }

  get _Title() {
    return this.tag('Content.Title');
  }
  get _Icon() {
    return this.tag('Content.Icon');
  }
  get _Stroke() {
    return this.tag('Stroke');
  }
}

export default withStyles(Button, styles);
