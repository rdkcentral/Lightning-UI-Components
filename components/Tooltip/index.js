import lng from 'wpe-lightning';
import {
  GRID,
  TYPESCALE,
  CORNER_RADIUS,
  getHexColor,
  COLORS_NEUTRAL,
  COLORS_TEXT
} from '../Styles';

import { RoundRect } from '../../utils';

const horizontalPadding = GRID.spacingIncrement * 2;

export default class Tooltip extends lng.Component {
  static _template() {
    return {
      alpha: 0,
      scale: 0.5,
      background: getHexColor(COLORS_NEUTRAL.light2),
      mountX: 0.5,
      x: w => w / 2,
      //TO DO: update to luminance vs dark drop shadows
      DropShadow: {
        zIndex: -1,
        x: w => w / 2,
        y: h => h / 2 + GRID.spacingIncrement * 2,
        mount: 0.5,
        color: getHexColor(COLORS_NEUTRAL.dark1, 80)
      },
      Text: {
        x: horizontalPadding,
        text: {
          ...TYPESCALE.footnote,
          textColor: getHexColor(COLORS_TEXT.dark),
          wordWrapWidth: 400,
          textAlign: 'center',
          verticalAlign: 'middle'
        }
      }
    };
  }

  set title(title) {
    this._Text.text = title;
    this._Text.on('txLoaded', () => {
      // set background size based on text size

      let backgroundW = this._Text.finalW + horizontalPadding * 2;
      let backgroundH =
        this._Text.finalH -
        (this._Text.finalH % TYPESCALE.footnote.lineHeight) / 2; // calculates even padding

      this.texture = lng.Tools.getRoundRect(
        RoundRect.getWidth(backgroundW),
        RoundRect.getHeight(backgroundH),
        CORNER_RADIUS.small
      );
      this._DropShadow.texture = lng.Tools.getShadowRect(
        backgroundW,
        backgroundH,
        CORNER_RADIUS.small,
        24
      );

      this.y =
        -backgroundH -
        (this.bottomMargin ? this.bottomMargin : GRID.spacingIncrement * 5);
    });
  }

  _focus() {
    clearTimeout(this._hideTimer);
    clearTimeout(this._showTimer);

    this.delayVisible
      ? (this._showTimer = setTimeout(() => {
          this.smooth = { alpha: 1, scale: 1 };
        }, this.delayVisible))
      : (this.smooth = { alpha: 1, scale: 1 });

    this._hideTimer = this.timeVisible
      ? setTimeout(() => {
          this._unfocus();
        }, this.timeVisible + this.delayVisible)
      : null;
  }

  _unfocus() {
    clearTimeout(this._hideTimer);
    clearTimeout(this._showTimer);

    this.smooth = { alpha: 0, scale: 0.5 };
  }

  get _Text() {
    return this.tag('Text');
  }
  get _DropShadow() {
    return this.tag('DropShadow');
  }
}
