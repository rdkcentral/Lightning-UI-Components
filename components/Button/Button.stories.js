import lng from 'wpe-lightning';

import Button from '.';
import withStyles from '../../mixins/withStyles';
import mdx from './Button.mdx';
import { action } from '@storybook/addon-actions';
import { boolean, number, withKnobs, text } from '@storybook/addon-knobs';
import icon from '../../assets/images/ic_lightning_white_32.png';

export default {
  title: 'Button',
  component: Button,
  decorators: [withKnobs],
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const Basic = () =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Button: {
          type: Button,
          title: text('Title', 'Button'),
          radius: number('Radius', 0, { min: 0, max: 20 }),
          onEnter: action('onEnter')
        }
      };
    }

    _init() {
      if (boolean('Icon', false)) {
        this.tag('Button').icon = { src: icon };
      }
      if (boolean('Stroke', false)) {
        this.tag('Button').stroke = { weight: 3, color: 0xff616161 };
      }
      if (boolean('Fixed', false)) {
        this.tag('Button').fixed = true;
      }
    }

    _getFocused() {
      if (boolean('Focused', false)) {
        return this.tag('Button');
      }
    }
  };

export const LongTitle = () =>
  class LongTitle extends lng.Component {
    static _template() {
      return {
        Button: {
          type: Button,
          title: 'This is a button with a really long title'
        }
      };
    }
    _getFocused() {
      if (boolean('Focused', false)) {
        return this.tag('Button');
      }
    }
  };

let orangeStyles = {
  background: {
    color: 0xffffd18c
  },
  icon: {
    color: 0xfffea73d
  },
  text: {
    color: 0xfffea73d,
    fontSize: 30
  },
  radius: 4,
  stroke: { weight: 3, color: 0xfffea73d },
  w: 200,
  h: 50,
  padding: 50,
  unfocus: {
    patch: function() {
      this.smooth = { scale: 1 };
    }
  },
  focus: {
    patch: function() {
      this.smooth = { scale: 1.1 };
    }
  }
};

export const WithTheme = () =>
  class WithTheme extends lng.Component {
    static _template() {
      return {
        Button: {
          type: withStyles(Button, orangeStyles),
          title: 'Button'
        }
      };
    }
    _getFocused() {
      if (boolean('Focused', false)) {
        return this.tag('Button');
      }
    }
  };

export const KeyHandling = () =>
  class KeyHandling extends lng.Component {
    static _template() {
      return {
        Button: {
          type: Button,
          title: 'Button',
          onEnter: button => (button.title = button.title + '!')
        }
      };
    }

    _getFocused() {
      return this.tag('Button');
    }
  };
