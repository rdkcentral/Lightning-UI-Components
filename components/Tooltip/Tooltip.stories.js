import lng from '@lightningjs/core';

import Tooltip from '.';
import mdx from './Tooltip.mdx';
import Button from '../Button';
import ActionButton, { styles as ACTION_BUTTON_STYLES } from '../ActionButton';

export default {
  title: 'Tooltip',
  args: {
    focused: false,
    title: 'Type Something',
    bottomMargin: 0,
    delayVisible: 0,
    timeVisible: 0
  },
  argTypes: {
    focused: { control: 'boolean' },
    title: { control: 'text' },
    bottomMargin: { control: 'number' },
    delayVisible: { control: 'number' },
    timeVisible: { control: 'number' }
  },
  parameters: {
    tag: 'Button.Tooltip',
    argActions: {
      focused: (isFocused, component) => {
        component._getFocused = isFocused
          ? () => component.tag('Button.Tooltip')
          : () => {};
        component._refocus();
      }
    },
    docs: {
      page: mdx
    }
  }
};

export const Basic = args =>
  class Basic extends lng.Component {
    static _template() {
      return {
        Button: {
          type: Button,
          x: 300,
          y: 100,
          title: 'Button',
          Tooltip: {
            type: Tooltip,
            title: args.title,
            bottomMargin: args.bottomMargin,
            delayVisible: args.delayVisible,
            timeVisible: args.timeVisible
          }
        }
      };
    }

    _getFocused() {
      if (args.focused) {
        return this.tag('Button.Tooltip');
      }
    }
  };

export const LongTitle = Basic.bind({});
LongTitle.args = {
  title: 'Type something with a long message here...'
};

export const WithActionButton = args =>
  class WithActionButton extends lng.Component {
    static _template() {
      return {
        Button: {
          type: ActionButton,
          x: 200,
          y: 100,
          title: 'Button',
          backgroundType: 'stroke',
          Tooltip: {
            type: Tooltip,
            title: args.title,
            bottomMargin: args.bottomMargin,
            delayVisible: args.delayVisible,
            timeVisible: args.timeVisible
          }
        }
      };
    }

    _getFocused() {
      if (args.focused) {
        return this.tag('Button.Tooltip');
      }
    }
  };
WithActionButton.args = {
  bottomMargin: 24
};
