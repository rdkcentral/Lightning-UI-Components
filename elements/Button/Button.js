/**
 * Button Component
 *
 * Generic Button Component that handles setting text icon and checkbox
 */
import { withExtensions } from '../../mixins';
import Surface from '../Surface';
import TextBox from '../TextBox';
import * as styles from './Button.styles';
import { measureTextWidth } from '../../utils';
import Row from '../../layout/Row';

class Button extends Surface {
  static get __componentName() {
    return 'Button';
  }

  static get __themeStyles() {
    return styles;
  }

  static get properties() {
    return ['fixed', 'justify', 'prefix', 'suffix', 'title'];
  }

  static get tags() {
    return [
      ...super.tags,
      'Content',
      { name: 'Title', path: 'Content.Title' },
      { name: 'Prefix', path: 'Content.Prefix' },
      { name: 'Suffix', path: 'Content.Suffix' }
    ];
  }

  static _template() {
    return {
      ...super._template(),
      Content: {
        mount: 0.5,
        x: w => w / 2,
        y: h => h / 2
      }
    };
  }

  _update() {
    this._updatePrefix();
    this._updateTitle();
    this._updateSuffix();
    this._updateAllPositioning();
    this._updateTruncation();
  }

  _updateAllPositioning() {
    this._updatePositions();
    this._updateContentDimensions();
    this._updateSurfaceDimensions();
    this._updateContentPosition();
    // delaying the super._update call so we update the surface sizing after the button has finished updating
    super._update();
  }

  $itemChanged() {
    this._updateAllPositioning();
  }

  _onTextBoxChanged() {
    this._updateAllPositioning();
  }

  _updatePositions() {
    if (this._hasPrefix) {
      this._Prefix.x = this._prefixX;
    }
    if (this._hasTitle) {
      this._Title.x = this._titleX;
    }
    if (this._hasSuffix) {
      this._Suffix.x = this._suffixX;
    }
  }

  _updatePrefix() {
    if (this.prefix) {
      let prefixPatch = {
        h: this.style.prefixH,
        itemSpacing: this.style.itemSpacing
      };
      if (!this._Prefix) {
        prefixPatch = {
          ...this._rowProps,
          ...prefixPatch
        };
      }
      this._Content.patch({ Prefix: prefixPatch });
      this._Prefix.items = this._addButtonProps(this.prefix);
    } else {
      this._Content.patch({ Prefix: undefined });
    }
  }

  _updateTitle() {
    if (this._hasTitle) {
      let titlePatch = {
        content: this.title,
        textStyle: {
          textColor: this.style.contentColor,
          ...this.style.textStyle
        }
      };
      if (!this._Title) {
        titlePatch = {
          type: TextBox,
          mountY: 0.5,
          y: h => h / 2,
          signals: {
            textBoxChanged: '_onTextBoxChanged'
          },
          ...titlePatch
        };
      }
      this._Content.patch({ Title: titlePatch });
    } else {
      this.patch({ Content: { Title: undefined } });
    }
  }

  _updateSuffix() {
    if (this.suffix) {
      let suffixPatch = {
        h: this.style.suffixH,
        itemSpacing: this.style.itemSpacing
      };
      if (!this._Suffix) {
        suffixPatch = {
          ...this._rowProps,
          ...suffixPatch
        };
      }
      this._Content.patch({ Suffix: suffixPatch });
      this._Suffix.items = this._addButtonProps(this.suffix);
    } else {
      this._Content.patch({ Suffix: undefined });
    }
  }

  _updateTruncation() {
    /**
     * needs to run before rest of content updates to prevent a bug causing text
     * to sometimes truncate and then un-truncate(with no width change) when
     * toggling fixed after toggling focus
     */
    if (this._Title) {
      this._Title.patch({
        wordWrap: this.fixed,
        wordWrapWidth: this.fixed ? this.w - this._paddingX : 0
      });
    }
  }

  _updateContentDimensions() {
    const contentDimensionsPatch = {};
    const y = this.h / 2;

    // reference transition targets to ensure we aren't checking against intermittent values
    if (this._Content.transition('w').targetValue !== this._contentW) {
      contentDimensionsPatch.w = this._contentW;
    }

    if (this._Content.y !== y) {
      contentDimensionsPatch.y = y;
    }

    if (Object.keys(contentDimensionsPatch).length > 0) {
      if (this._smooth) {
        this._Content.smooth = contentDimensionsPatch;
      } else {
        this._Content.patch(contentDimensionsPatch);
      }
    }
  }

  _updateContentPosition() {
    let mountX;
    let x;

    switch (this.justify) {
      case 'left':
        mountX = 0;
        x = this._paddingLeft;
        break;
      case 'right':
        mountX = 1;
        x = this.w - this._paddingRight;
        break;
      case 'center':
      default:
        mountX = 0.5;
        x = this.w / 2;
        break;
    }

    this._Content.patch({ mountX, x });
  }

  _updateSurfaceDimensions() {
    let newWidth = this.w;
    if (this.fixed) {
      newWidth = this._w;
    } else {
      // if no title, ignore minWidth and use prefix/suffix width
      newWidth =
        !this._hasTitle && (this._hasPrefix || this._hasSuffix)
          ? this._contentW + this._paddingX
          : Math.max(this._contentW + this._paddingX, this.style.minWidth);
    }

    if (newWidth !== this.w) {
      this.w = newWidth;
      this.fireAncestors('$itemChanged');
    }
  }

  _addButtonProps(arr) {
    const items = Array.isArray(arr) ? arr : [arr];
    return items.map(item => ({
      centerInParent: true,
      palette: this.palette,
      mode: this.mode,
      ...item,
      style: {
        color: this.style.contentColor,
        ...item.style
      }
    }));
  }

  _getJustify() {
    return !!this._justify ? this._justify : this.style.justify;
  }

  get _hasPrefix() {
    return this._Prefix && Array.isArray(this.prefix)
      ? this.prefix.length
      : this.prefix;
  }

  get _prefixW() {
    return this._hasPrefix ? this._Prefix.w : 0;
  }

  get _prefixX() {
    return 0;
  }

  get _hasTitle() {
    return !!this.title;
  }

  get _titleW() {
    // get the pre-calculated text width of the title to solve some race condition bugs
    if (
      this._hasTitle &&
      this._Title &&
      this._Title._Text &&
      this._Title.visible
    ) {
      if (this.fixed) {
        return this._Title.w;
      } else {
        return measureTextWidth(this._Title._Text.text);
      }
    }
    return 0;
  }

  get _titleX() {
    return this._hasPrefix ? this._prefixW + this.style.titlePadding : 0;
  }

  get _hasSuffix() {
    return this._Suffix && Array.isArray(this.suffix)
      ? this.suffix.length
      : this.suffix;
  }

  get _suffixW() {
    return this._hasSuffix ? this._Suffix.w : 0;
  }

  get _suffixX() {
    if (this._hasTitle) {
      return this._titleW + this._Title.x + this.style.titlePadding;
    } else if (this._hasPrefix) {
      return this._prefixW + this.style.itemSpacing;
    }
    return 0;
  }

  get _contentW() {
    if (this._hasSuffix) {
      return this._suffixX + this._suffixW;
    } else if (this._hasTitle && this._Title.visible) {
      return this._titleX + this._titleW;
    } else if (this._hasPrefix) {
      return this._prefixX + this._prefixW;
    }
    return 0;
  }

  get _rowProps() {
    return {
      type: Row,
      mountY: 0.5,
      autoResizeWidth: true
    };
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    if (this._announce) {
      return this._announce;
    }

    // TODO - Localization?
    // Do we need a locale file with
    // component translations?
    return this.title + ', Button';
  }

  get _paddingX() {
    return this._paddingLeft + this._paddingRight;
  }

  get _paddingLeft() {
    return this._hasTitle ? this.style.paddingX : this.style.paddingXNoTitle;
  }

  get _paddingRight() {
    return this._hasTitle ? this.style.paddingX : this.style.paddingXNoTitle;
  }
}

export default withExtensions(Button);
