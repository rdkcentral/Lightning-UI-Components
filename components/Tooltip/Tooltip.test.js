'use strict';

import TestUtils from '../lightning-test-utils';
import Tooltip from '.';
import Button from '../Button';

const createTooltip = TestUtils.makeCreateComponent(
  Tooltip,
  {},
  { focused: false }
);
const createButton = TestUtils.makeCreateComponent(Button);

describe('Tooltip', () => {
  let tooltip, testRenderer;

  beforeEach(() => {
    [tooltip, testRenderer] = createTooltip({ title: 'Tooltip' });
  });
  afterEach(() => {
    tooltip = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  describe('visibility', () => {
    it('should be visible when focused', () => {
      tooltip._focus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);
    });
    it('should be hidden when unfocused', () => {
      tooltip._unfocus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(0);
    });
    it('should show after delay if delayVisible is set', () => {
      jest.useFakeTimers();
      expect(tooltip.alpha).toEqual(0);

      // set delayVisible
      tooltip.delayVisible = 5000;

      // should remain hidden
      tooltip._focus();
      testRenderer.update();
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(tooltip.alpha).toEqual(0);

      // should show after timer runs
      jest.runAllTimers();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);
    });
    it('should hide after timer ends if timeVisible is set', () => {
      jest.useFakeTimers();

      // timer not run unless focused
      expect(setTimeout).toHaveBeenCalledTimes(0);

      // show tooltip on focus
      tooltip._focus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);

      // check that no timer has been called and tooltip remains visible
      jest.runAllTimers();
      expect(setTimeout).toHaveBeenCalledTimes(0);
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);

      // set timeVisilbe
      tooltip.timeVisible = 5000;

      // timer not run unless focused
      expect(setTimeout).toHaveBeenCalledTimes(0);

      tooltip._focus();
      testRenderer.update();
      expect(tooltip.alpha).toEqual(1);

      // timer started on focus and hide after timer is called
      jest.runAllTimers();
      expect(setTimeout).toHaveBeenCalledTimes(1);
      testRenderer.update();
      expect(tooltip.alpha).toEqual(0);
    });
  });

  describe('position', () => {
    let button, buttonRenderer;
    beforeEach(() => {
      tooltip._focus();

      [button, buttonRenderer] = createButton({ title: 'Button' });
      buttonRenderer.update();
    });
    it('should be horizontally centered', () => {});
    it('should default to 40px above parent', done => {
      tooltip._Text.on('txLoaded', () => {
        let buffer = 2;
        expect(tooltip.y).toEqual(-tooltip.finalH - 40 + buffer);

        done();
      });
    });

    it('should change when "bottomMargin" is set', done => {
      let closer_tooltip, closer_testRenderer;
      [closer_tooltip, closer_testRenderer] = createTooltip({
        title: 'Tooltip',
        bottomMargin: 20
      });
      closer_tooltip._focus();

      tooltip._Text.on('txLoaded', () => {
        closer_tooltip._Text.on('txLoaded', () => {
          expect(closer_tooltip.y).toEqual(tooltip.y + 20);
          done();
        });
      });
    });
  });

  describe('dimensions', () => {
    // NOTE: the font doesn't text size and calculated tooltip size is not accurate

    beforeEach(() => {
      tooltip._focus();
    });

    it('should be slightly larger than text dimensions', done => {
      tooltip._Text.on('txLoaded', () => {
        let textWidth = tooltip._Text.finalW;
        let textHeight = tooltip._Text.finalH;
        let diffW = Math.abs(tooltip._Text.finalW - textWidth);

        expect(diffW).toBeLessThan(5);
        expect(tooltip._DropShadow.finalW).toBeGreaterThan(textWidth);

        let diffH = Math.abs(tooltip.finalH - textHeight);
        expect(diffH).toBeLessThan(5);
        expect(tooltip._DropShadow.finalH).toBeGreaterThan(textHeight);

        done();
      });
    });

    it('should change width when hint message text changes', done => {
      let short_tooltip, short_testRenderer;
      [short_tooltip, short_testRenderer] = createTooltip({ title: 'Short' });
      short_tooltip._focus();

      let long_tooltip, long_testRenderer;
      [long_tooltip, long_testRenderer] = createTooltip({
        title: 'Long Tooltip'
      });
      long_tooltip._focus();

      tooltip._Text.on('txLoaded', () => {
        let controlW = tooltip.finalW;
        let controlShadowW = tooltip._DropShadow.finalW;

        short_tooltip._Text.on('txLoaded', () => {
          expect(short_tooltip.finalW).toBeLessThan(controlW);
          expect(short_tooltip._DropShadow.finalW).toBeLessThan(controlShadowW);

          long_tooltip._Text.on('txLoaded', () => {
            expect(long_tooltip.finalW).toBeGreaterThan(controlW);
            expect(long_tooltip._DropShadow.finalW).toBeGreaterThan(
              controlShadowW
            );

            done();
          });
        });
      });
    });

    it('should change height and wrap text when hint message is long', done => {
      let twoLine_tooltip, twoLine_testRenderer;
      [twoLine_tooltip, twoLine_testRenderer] = createTooltip({
        title:
          'This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...This text should render in two lines...'
      });
      twoLine_tooltip._focus();

      tooltip._Text.on('txLoaded', () => {
        twoLine_tooltip._Text.on('txLoaded', () => {
          let textHeight = twoLine_tooltip._Text.finalH;
          let diffH = Math.abs(twoLine_tooltip.finalH - textHeight);
          expect(diffH).toBeLessThan(5);
          expect(twoLine_tooltip._DropShadow.finalH).toBeGreaterThan(
            textHeight
          );

          expect(twoLine_tooltip.finalH).toBeGreaterThan(tooltip.finalH);
          expect(twoLine_tooltip._Text.finalH).toBeGreaterThan(
            tooltip._Text.finalH
          );
          expect(twoLine_tooltip._DropShadow.finalH).toBeGreaterThan(
            tooltip._DropShadow.finalH
          );

          done();
        });
      });
    });
  });
});
