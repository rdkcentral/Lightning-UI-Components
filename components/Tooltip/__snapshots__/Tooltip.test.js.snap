// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders 1`] = `
Object {
  "Component": Object {
    "DropShadow": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": "cc000000",
      "enabled": false,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "id": 3,
      "isComponent": undefined,
      "mount": 0.5,
      "mountX": 0.5,
      "mountY": 0.5,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "DropShadow",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "DropShadow",
      ],
      "visible": true,
      "w": 0,
      "x": [Function],
      "y": [Function],
      "zIndex": -1,
    },
    "Text": Object {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294967295,
      "enabled": false,
      "flex": false,
      "flexItem": false,
      "h": 0,
      "id": 4,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Text",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": Array [
        "Text",
      ],
      "texture": Object {
        "fontFace": "XfinityStandardBold",
        "fontSize": 24,
        "lineHeight": 32,
        "maxLinesSuffix": "...",
        "text": "Tooltip",
        "textAlign": "center",
        "textColor": 4278650631,
        "textOverflow": "ellipsis",
        "type": "TextTexture",
        "wordWrapWidth": 400,
      },
      "visible": true,
      "w": 0,
      "x": 16,
      "y": 0,
      "zIndex": 0,
    },
  },
}
`;
